### DISCUSSION TITLE
Setting a static body dynamic with the newest version

### DISCUSSION THREAD

**Discussion by ia97lies** [dyn4j/dyn4j#280] [General]
With the version 4.x.x I could set a body from static to danymic by just setting the mass to Normal like this:
        body.setMassType(MassType.NORMAL);

But this seems not to work anymore, any suggestions how this is usually done? I need that for my falling blocks which I trigger when you are underneath or walk over them.

**Comment by wnbittle** [comment]
That should still work. You are calling the .setMass method earlier correct?

Could you share some code?  I can test this myself in few hours.

**Reply by ia97lies** [reply]
I don't have it isolated but it's basically this

First I do this
```
        var bb = getObject();
        var body = bb.getBody();
        body.storeMass(body.getMass());
        getObject().getBody().setMassType(MassType.INFINITE);
        getObject().getBody().setLinearVelocity(0, 0);
```

to set it static

later I do this
```
        var bb = getObject();
        var body = bb.getBody();
        var mass = body.getStoredMass();
        body.setMassType(MassType.NORMAL);
        body.setMass(mass);
        EntityData ed = bb.getEntityData();
        body.setAngularVelocity(0);
        body.getTransform().setRotation(0);
        ed.setComponents(ed.createEntity(),
                new Level(false),
                new Force(force.clone()),
                new Link(body.getEntiyId()),
                new Decay(dt));
        if (torque != 0) {
            ed.setComponents(ed.createEntity(),
                    new Level(false),
                    new Torque(torque),
                    new Link(body.getEntiyId()),
                    new Decay(dt));
        }
```

To set it dynamic add a bit force to spicy it up and as well some torque  via my entity component system.

And now looking at it, I probably don't even have to store the original mass when I set the type to Infinite.... But anyway, this works with 4.2.1 and does not with any 5.x.x version.

**Comment by wnbittle** [comment]
Yep, I can confirm.  For a quick fix, also wake the body up by `body.setAtRest(false);`.  I didn't see any change that would affect it yet - thanks for specifying which version of 4.x, that should help me track it down.

**Reply by ia97lies** [reply]
I tried already setAtRest(false) before, but that didn't help.

**Comment by wnbittle** [comment]
It could be this [The getMass and getType methods should be consistent](https://github.com/dyn4j/dyn4j/issues/235).

```diff
public void setType(MassType type) {
  if (type == null) throw new NullPointerException(Messages.getString("geometry.mass.nullMassType"));
  
+  // don't allow incorrect mass type + mass/inertia combinations
+  if (type == MassType.NORMAL && (this.mass <= 0.0 || this.inertia <= 0.0)) {
+    return;
+  } else if (type == MassType.FIXED_LINEAR_VELOCITY && (this.inertia <= 0.0)) {
+    return;
+  } else if (type == MassType.FIXED_ANGULAR_VELOCITY && (this.mass <= 0.0)) {
+    return;
+  }
  
  this.type = type;
}
```
The goal with this change was to make sure that you can't set something to `NORMAL`, but it still have mass/inertia of 0 since that would represent a different mass type.

The goal with the MassType feature was to allow you to set the mass of a body to what it would be if it wasn't some weird type.  Then, as needed, you could toggle the mass type without having to recalculate the mass.  For example: https://github.com/dyn4j/dyn4j/blob/master/src/main/java/org/dyn4j/geometry/Mass.java#L358

Can you show how you create the body and set the mass initially?

**Reply by ia97lies** [reply]
In this case I do this

```
        //body.setBullet(true);
        body.getFixture(0).setRestitution(0);
        body.getFixture(0).setFriction(0.5);
        body.getFixture(0).setFilter(new FallingBlockFilter());
        Mass mass = new Mass(new Vector2(0, 0), shape.getRadius() * 20, 0);
        body.setMass(mass);
        body.setAngularDamping(100);
        body.setAtRestDetectionEnabled(true);
```

**Reply by ia97lies** [reply]
When I hopped in setType(...) I didn't see those checks I use 5.0.1

**Reply by wnbittle** [reply]
I see.  In this case you'd need to use `MassType.FIXED_ANGULAR_VELOCITY` and it should work correctly.

The change came in 4.2.2 so I would expect it to be in every 5.x version.

**Reply by ia97lies** [reply]
Bit I have one without angular damping 100 that also doesn’t work. 

And I did the mass restoring now before I set the mass type but doesn’t work either, that should circumvent your new if testing. 

Anyway I currently moved back to 4.2.1 bc this one oc the main things of my game. And your right I tried a while ago 4.2.2 and it didn’t work either.

**Reply by ia97lies** [reply]
out of curiosity what for is MassType.FIXED_ANGULAR_VELOCITY exactly?

**Reply by ia97lies** [reply]
Uuuh the innerta == 0.0 is probably my problem, right? What is a good value for that anyway?

**Reply by ia97lies** [reply]
And fixed! Thank you for your help. My code is clearer now and I understand these flags better :D Learned something!

**Comment by ia97lies** [comment]
One more thing to add, if I use FIXED_ANGULAR_VELOCITY it sometimes get extremely weird fast rotations when it collides with something static. I probably use it wrong.... But the angular damping to a high number doesn't have this flicker/wild rotations. So I'm not sure what to do.
One thing to add I have a collider listener and there I remove the colliding static blocks, but seems the collision already messed the angular rotation logic and it is then in a funny rotated position and doesn't look right... It's very tricky :)

**Reply by ia97lies** [reply]
And solved this as well. Thanks for listening, classical rubber duck debugging ;) I had to store the mass type if I set it to INFINITY so I can restore it if I set it dynamic again, it doesn't work well with inerta == 0.0 in general I guess. So I hope I fixed all my issues with this. Thanks again!

**Comment by wnbittle** [comment]
Nice! sorry for not getting back on your other comments here earlier.

```java
Mass mass = new Mass(new Vector2(0, 0), shape.getRadius() * 20, 0);
```
Yep you got it - the third argument is the _inertia_ which is a measure of how hard it is to change the angular velocity of body.  In this case, you've set it to zero which is the same as `FIXED_ANGULAR_VELOCITY`.  It's called this because NOTHING can _change_ the body's angular velocity.  In most cases the angular velocity will be zero, but you could create a body with 0 inertia that is rotating.  Imagine a slowly spinning (fixed speed) platform that falls linearly when jumped on by the player.

Question for you - any reason why you aren't letting dyn4j compute the mass/inertia from the fixtures of the body via `setMass(MassType)`?

> One more thing to add, if I use `FIXED_ANGULAR_VELOCITY` it sometimes get extremely weird

Yeah, `FIXED_ANGULAR_VELOCITY`,  `FIXED_LINEAR_VELOCITY` and `INFINITE` will definitely act weird when interacting with each other.  

If you think about it from a real world perspective, these concepts don't really exist.  They're all contrived options to help simulate things in the real world in a stable way.  For example, let's say you want a platform that a player can jump on that spins freely but can never move linearly.  You could use a RevoluteJoint to keep that platform in place, but the joints can only handle so much before they explode and can no longer be solved (we have finite precision, nor can we solve at the atomic level the interactions).  Instead, you could set the platform body to `FIXED_LINEAR_VELOCITY` so that it will _never_ move from the pivot point, but can still rotate freely.

Take another example, let body1 be `INFINITE` and body2 be `FIXED_ANGULAR_VELOCITY`.  Imagine body1 is a long rectangular "floor".  Imagine body2 is a rectangle that's falling and is rotated at 45 degrees.  In the real world when body2 hits body1, one end of body2 would be stopped and the other end of body2 would continue to fall.  BUT, since we said body2 can't rotate, it would just stop completely with one end of body2 "resting" in the air like this:

```
  \
   \
    \
------------------------
```

Not very plausible for the real world, but that's what we've told the simulation to do.  For this reason, I don't usually recommend using `FIXED_LINEAR_VELOCITY` or `FIXED_ANGULAR_VELOCITY` unless you really need them.

**Reply by ia97lies** [reply]
Just I use this for my action platformer bc it looks cool. The blocks with 20 times radius are very heavy (concrete and iron) I don’t want to move them around accidentally. As well I have a 5 times higher gravity than earth to make jumps look snappy and fast with a realistic touch. Another pain point is collision listener which should smash bodes which are in between but the collision does stop even extremely havy hammers and I wish it wouldn’t 😂 so there are more tweaks to get the benefit of cool looking fragments flying around when a heavy duty block lands, super nice smashing blocks and more. But I tweak to the max. Sometimes I get weird behavior I rather don’t wand and hopped the engine could be more “reasonable” 😂 it sometimes is in a stated where collision don’t work bc of some weird behavior before with staic blocks 😅
